working/0110-incremental-sound-nnbd/roadmap.md

```diff
@@ -0,0 +1,874 @@
+
+On the second point, arrays are likely to be much less of an issue for Dart,
+since the structure of the `List` API encourages incremental building, whereas
+the core `C#` `Array` type does not support this.  Default constructors may be a
+place where some amount of Dart code will need to be modified, but Dart
+programming style encourages objects that are initialized in
+constructors, and provides good support for doing so.
+
+This leaves the third issue.  On the face of it, this flies in the face of the
+experience of other languages which have simply chosen to take the sound route
+on this with apparently no repercussions (Kotlin programmers seem to live fine
+with this, for example).  It's possible that their sense was that this made the
+migration story of large amounts of existing C# code easier, or that they simply
+felt that since they were unsound elsewhere, they might as well be here.  The
+migration story may come into play for Dart, but the scale of the problem is
+much smaller.
+
+#### [Java](https://blogs.oracle.com/java-platform-group/java-8s-new-type-annotations)
+
+Java provides nullability annotations that can be used by external tools to
+provide
+[sound or near-sound checking](https://checkerframework.org/manual/#nullness-checker).
+The annotations are not interpreted with the language at all.  The Java
+documentation
+[suggests](https://blogs.oracle.com/java-platform-group/java-8s-new-type-annotations) not
+relying on nullability annotations, saying "Optional Type Annotations are not a
+substitute for runtime validation".
+
+#### [Typescript](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html)
+
+Typescript is of course unsound by design.  In addition to the core unsoundness
+of the type system, it specifically allows unsound nullable promotion of dotted
+paths.
+
+```typescript
+class Ref<T> {
+  v : T;
+  constructor(_v : T) {this.v = _v;}
+}
+
+function main() {
+	var rn: Ref<Ref<number | null>> = new Ref(new Ref(3));
+	var rn2 = rn;
+
+	if (rn.v.v != null) {
+		rn2.v.v = null;
+		var x : number = rn.v.v;  // Assigns null to a non-nullable type
+	}
+}
+```
+
+As with other languages, Typescript provides an assertion operator `e!.foo`
+which asserts that `e` is non-null.  Unlike Swift and Kotlin, this operator is
+not dynamically checked.
+
+Typescript also provides an option to do definite assignment analysis in
+constructors to help enforce proper object construction.
+
+```typescript
+class IntRef {
+  v : number; // Error, not definitely assigned in constructor
+	constructor(_v: number) { }
+}
+```
+
+As with Kotlin and Swift, Typescript does not seem to have skimped on the static
+analysis work, but instead has invested fairly heavily in providing robust
+checking.  As with C# however, Typescript does seem to have chosen to be unsound
+in a few places out of choice rather than necessity, to relieve the programmer
+of some of the burden of proving safety to the type checker.
```

Comment:
Probably worth pointing out that even if they were sound, they wouldn't be able to extract any performance benefit from it.

Reply:
done.